File: src/mini_whisper/pricing.py
"""Dictation cost maths and menu usage-row formatting.

Rates are editable defaults; individual constants can be overridden via the
`pricing_overrides` dict in config.json. Pure module — shared by controller
and app without circular imports.
"""

PER_MINUTE = {
    "openai_realtime": 0.017,
    "elevenlabs": 0.0065,
    "speechmatics": 0.0067,
    "on_device": 0.0,
}

PER_MTOK = {
    "gpt-4o-mini-transcribe_in": 1.25,
    "gpt-4o-mini-transcribe_out": 5.00,
    "gpt-4o-mini_in": 0.15,
    "gpt-4o-mini_out": 0.60,
}

# Engine names (StreamingEngine.name) that differ from their rate-table key
_ENGINE_RATE_KEYS = {"openai": "openai_realtime"}


def _rate(table: dict, key: str, overrides: dict | None) -> float:
    if overrides and key in overrides:
        return float(overrides[key])
    return table.get(key, 0.0)


def dictation_cost(
    engine_name: str | None,
    seconds: float,
    token_usage_by_model: dict,
    overrides: dict | None = None,
) -> float:
    """Dollar cost of one dictation.

    engine_name: streaming engine billed for `seconds` of audio (None = batch only).
    token_usage_by_model: {model: {"input_tokens": n, "output_tokens": n}}.
    """
    cost = 0.0
    if engine_name is not None:
        rate_key = _ENGINE_RATE_KEYS.get(engine_name, engine_name)
        cost += (seconds / 60.0) * _rate(PER_MINUTE, rate_key, overrides)
    for model, tokens in token_usage_by_model.items():
        cost += (tokens.get("input_tokens", 0) / 1_000_000) * _rate(PER_MTOK, f"{model}_in", overrides)
        cost += (tokens.get("output_tokens", 0) / 1_000_000) * _rate(PER_MTOK, f"{model}_out", overrides)
    return round(cost, 6)


def _fmt_tokens(n) -> str:
    n = int(n)
    return f"{n / 1000:.1f}k" if n >= 1000 else str(n)


def format_usage_rows(today: dict, month_cost: float) -> tuple[str, str]:
    """Render the two menu rows from today's usage entry and month-to-date cost."""
    minutes = int(sum(today.get("streamed_seconds", {}).values()) // 60)
    today_row = (
        f"Today: {_fmt_tokens(today.get('input_tokens', 0))}/"
        f"{_fmt_tokens(today.get('output_tokens', 0))} tok"
        f" · {minutes}m · ${today.get('cost_usd', 0.0):.2f}"
    )
    month_row = f"Month: ${month_cost:.2f}"
    return today_row, month_row

File: Packages/MiniWhisperCore/Sources/MWUsage/Pricing.swift
import Foundation
import MWConfig
import MWSupport

/// Dictation cost maths and the two menu rows (F35, F31), a port of
/// `../mini-whisper-py/src/mini_whisper/pricing.py` with the added `speech_analyzer` rate.
/// Every constant is overridable through `config.json`'s `pricing_overrides`.
public enum Pricing {
    public static let perMinute: [String: Double] = [
        "openai_realtime": 0.017,
        "elevenlabs": 0.0065,
        "speechmatics": 0.0067,
        "on_device": 0.0,
        "speech_analyzer": 0.0,
    ]

    public static let perMTok: [String: Double] = [
        "gpt-4o-mini-transcribe_in": 1.25,
        "gpt-4o-mini-transcribe_out": 5.00,
        "gpt-4o-mini_in": 0.15,
        "gpt-4o-mini_out": 0.60,
    ]

    /// Engine names whose rate-table key differs from the name.
    private static let engineRateKeys: [EngineName: String] = [.openai: "openai_realtime"]

    public static func rate(
        _ table: [String: Double],
        key: String,
        overrides: [String: Double] = [:]
    ) -> Double {
        overrides[key] ?? table[key] ?? 0
    }

    /// `engine` is the streaming engine billed for `seconds` (nil = batch only);
    /// `tokensByModel` bills the OpenAI models the dictation used.
    public static func dictationCost(
        engine: EngineName?,
        seconds: TimeInterval,
        tokensByModel: [String: TokenUsage],
        overrides: [String: Double] = [:]
    ) -> Double {
        var cost = 0.0
        if let engine {
            let key = engineRateKeys[engine] ?? engine.rawValue
            cost += (seconds / 60) * rate(perMinute, key: key, overrides: overrides)
        }
        // Sorted so the floating-point sum does not depend on dictionary order.
        for model in tokensByModel.keys.sorted() {
            guard let tokens = tokensByModel[model] else { continue }
            cost += Double(tokens.inputTokens) / 1_000_000
                * rate(perMTok, key: "\(model)_in", overrides: overrides)
            cost += Double(tokens.outputTokens) / 1_000_000
                * rate(perMTok, key: "\(model)_out", overrides: overrides)
        }
        return (cost * 1_000_000).rounded() / 1_000_000
    }

    /// The two menu rows (F31): `Today: 1.2k/340 tok · 3m · $0.12` and `Month: $1.50`.
    public static func formatUsageRows(
        today: DayEntry,
        monthCost: Double
    ) -> (today: String, month: String) {
        let minutes = Int((today.streamedSeconds.values.reduce(0, +) / 60).rounded(.down))
        let todayRow = "Today: \(tokens(today.inputTokens))/\(tokens(today.outputTokens)) tok"
            + " · \(minutes)m · \(dollars(today.costUSD))"
        return (todayRow, "Month: \(dollars(monthCost))")
    }

    private static func tokens(_ count: Int) -> String {
        count >= 1000 ? String(format: "%.1fk", Double(count) / 1000) : String(count)
    }

    /// `$1.50`, or `<$0.01` when the amount is real but too small to show at this
    /// precision — a cleanup call costs about $0.00004, and `$0.000` reads as free.
    public static func dollars(_ amount: Double, decimals: Int = 2) -> String {
        let smallest = pow(10, -Double(decimals))
        if amount > 0, amount < smallest / 2 {
            return String(format: "<$%.\(decimals)f", smallest)
        }
        return String(format: "$%.\(decimals)f", amount)
    }
}
